src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -0,0 +1,224 @@
+pub type Port = usize;
+
+#[derive(Clone)]
+pub struct NailgunPool {
+    processes: Arc<Mutex<NailgunProcessMap>>,
+}
+
+impl NailgunPool {
+    pub fn new() -> Self {
+        NailgunPool {
+            processes: Arc::new(Mutex::new(NailgunProcessMap::new())),
+        }
+    }
+
+    ///
+    /// Given a `NailgunProcessName` and a `ExecuteProcessRequest` configuration,
+    /// return a port of a nailgun server running under that name and configuration.
+    ///
+    /// If the server is not running, or if it's running with a different configuration,
+    /// this code will start a new server as a side effect.
+    ///
+    pub fn connect(
+        &self,
+        name: NailgunProcessName,
+        startup_options: ExecuteProcessRequest,
+        workdir_path: &PathBuf,
+        nailgun_req_digest: Digest,
+    ) -> Result<Port, String> {
+        trace!("Locking nailgun process pool so that only one can be connecting at a time.");
+        let mut processes = self.processes.lock();
+
+        let maybe_process = processes.get_mut(&name);
+        let connection_result = if let Some(process) = maybe_process {
+            // Clone some fields that we need for later
+            let (process_name, process_fingerprint, process_port) =
+                (process.name.clone(), process.fingerprint.clone(), process.port);
+
+            debug!(
+                "Checking if nailgun server {} is still alive at port {}...",
+                &process_name, process_port
+            );
+
+            // If the process is in the map, check if it's alive using the handle.
+            let status = {
+                process
+                    .handle
+                    .lock()
+                    .try_wait()
+                    .map_err(|e| format!("Error getting the process status! {}", e))
+                    .clone()
+            };
+            match status {
+                Ok(None) => { // Process hasn't exited yet
+                    debug!("Found nailgun process {}, with fingerprint {:?}",
+                           &name, process_fingerprint);
+                    if nailgun_req_digest == process_fingerprint {
+                        debug!("The fingerprint of the running nailgun {:?} matches the requested fingerprint {:?}. Connecting to existing server.",
+                               nailgun_req_digest, process_fingerprint);
+                        Ok(process_port)
+                    } else {
+                        // The running process doesn't coincide with the options we want.
+                        // Restart it.
+                        debug!("The options for process {} are different to the startup_options! \n Startup Options: {:?}\n Process Cmd: {:?}",
+                               &process_name, startup_options, process_fingerprint
+                        );
+                        self.start_new_nailgun(&mut *processes, name, startup_options, workdir_path, nailgun_req_digest)
```

Comment:
This kills the existing nailgun, right? Do we have a form of synchronisation here to make sure no process is _using_ the nailgun we're aggressively killing?

Reply:
After syncing online, addressed in https://github.com/pantsbuild/pants/pull/8371/commits/c0ec22c374a1b281be3be376cb2a9aa0eef3f37b